File: interator.py
from collections.abc import Iterator, Iterable

class Cancion:
    def __init__(self, titulo, artista):
        self.titulo = titulo
        self.artista = artista

    def __str__(self):
        return f'"{self.titulo}" de {self.artista}'

class ListaReproduccionIterator(Iterator):
    def __init__(self, canciones):
        self.canciones = canciones
        self.indice = 0

    def __next__(self):
        if self.indice >= len(self.canciones):
            raise StopIteration
        cancion = self.canciones[self.indice]
        self.indice += 1
        return cancion

class ListaReproduccion(Iterable):
    def __init__(self):
        self.canciones = []

    def agregar_cancion(self, cancion):
        self.canciones.append(cancion)
    
    def __iter__(self):
        return ListaReproduccionIterator(self.canciones)

# Crear lista de reproducción y agregar canciones
playlist = ListaReproduccion()

playlist.agregar_cancion(Cancion("Bohemian Rhapsody", "Queen"))
playlist.agregar_cancion(Cancion("Smells Like Teen Spirit", "Nirvana"))
playlist.agregar_cancion(Cancion("Hotel California", "Eagles"))

# Iterar sobre la lista de reproducción
for cancion in playlist:
    print(cancion)

File: java/interator.java
import java.util.Iterator;
import java.util.ArrayList;
import java.util.List;
import java.util.NoSuchElementException;

// Clase que representa una canción
class Cancion {
    private String titulo;
    private String artista;

    public Cancion(String titulo, String artista) {
        this.titulo = titulo;
        this.artista = artista;
    }

    @Override
    public String toString() {
        return "\"" + titulo + "\" de " + artista;
    }
}

// Iterador para la lista de reproducción
class ListaReproduccionIterator implements Iterator<Cancion> {
    private List<Cancion> canciones;
    private int indice;

    public ListaReproduccionIterator(List<Cancion> canciones) {
        this.canciones = canciones;
        this.indice = 0;
    }

    @Override
    public boolean hasNext() {
        return indice < canciones.size();
    }

    @Override
    public Cancion next() {
        if (!hasNext()) {
            throw new NoSuchElementException();
        }
        Cancion cancion = canciones.get(indice);
        indice++;
        return cancion;
    }
}

// Clase que representa una lista de reproducción (Iterable)
class ListaReproduccion implements Iterable<Cancion> {
    private List<Cancion> canciones;

    public ListaReproduccion() {
        canciones = new ArrayList<>();
    }

    public void agregarCancion(Cancion cancion) {
        canciones.add(cancion);
    }

    @Override
    public Iterator<Cancion> iterator() {
        return new ListaReproduccionIterator(canciones);
    }
}

// Clase pública con el método main
public class interator {
    public static void main(String[] args) {
        ListaReproduccion playlist = new ListaReproduccion();
        playlist.agregarCancion(new Cancion("Bohemian Rhapsody", "Queen"));
        playlist.agregarCancion(new Cancion("Smells Like Teen Spirit", "Nirvana"));
        playlist.agregarCancion(new Cancion("Hotel California", "Eagles"));

        // Iterar sobre la lista de reproducción e imprimir cada canción
        for (Cancion cancion : playlist) {
            System.out.println(cancion);
        }
    }
}
